source/common/router/router.cc

```diff
@@ -470,6 +476,40 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
   return Http::FilterTrailersStatus::StopIteration;
 }
 
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  Http::MetadataMap metadata;
+  metadata.insert(metadata_map.begin(), metadata_map.end());
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
+
+  if (upstream_requests_.front() == nullptr) {
+    ENVOY_STREAM_LOG(trace,
```

Comment:
My goal in eagerly creating the upstream request was to not have to queue the metadata in two places.  Can we handle route/pool failure here as well, so we don't have to buffer?

Reply:
Removed upstream_requests_ creation because headers should always arrive before metadata, :D Please see the explanation in the comments below.